File: src/geometry_triangle_congruence_medium.py
"""
geometry - triangle_congruence (medium)
Generated: 2026-02-22T04:06:40.688032
"""

from problem_utils import *
from svg_utils import Diagram

def generate():
    problem_type = randint(1, 4)
    
    if problem_type == 1:
        # SSS Congruence - given three sides, determine if triangles are congruent
        side1 = randint(3, 8)
        side2 = randint(4, 9)
        side3 = randint(5, 10)
        
        # Make sure triangle inequality holds
        while side1 + side2 <= side3 or side1 + side3 <= side2 or side2 + side3 <= side1:
            side3 = randint(5, 10)
        
        # Second triangle - sometimes congruent, sometimes not
        if randint(0, 1) == 0:
            # Make congruent
            sides2 = [side1, side2, side3]
            from random import shuffle
            shuffle(sides2)
            ans = True
        else:
            # Make not congruent by changing one side
            sides2 = [side1, side2, side3 + randint(1, 3)]
            ans = False
        
        d = Diagram(width=400, height=200, padding=20)
        
        # First triangle
        p1 = (0, 0)
        p2 = (side1 * 15, 0)
        p3 = (side1 * 7.5, side2 * 12)
        d.polygon([p1, p2, p3], labels=['A', 'B', 'C'])
        d.segment_label(p1, p2, str(side1))
        d.segment_label(p2, p3, str(side2))
        d.segment_label(p3, p1, str(side3))
        
        # Second triangle
        q1 = (side1 * 20, 0)
        q2 = (side1 * 20 + sides2[0] * 15, 0)
        q3 = (side1 * 20 + sides2[0] * 7.5, sides2[1] * 12)
        d.polygon([q1, q2, q3], labels=['D', 'E', 'F'])
        d.segment_label(q1, q2, str(sides2[0]))
        d.segment_label(q2, q3, str(sides2[1]))
        d.segment_label(q3, q1, str(sides2[2]))
        
        return problem(
            question=f"Are triangles ABC and DEF congruent? Triangle ABC has sides {side1}, {side2}, {side3}. Triangle DEF has sides {sides2[0]}, {sides2[1]}, {sides2[2]}.",
            answer=ans,
            difficulty=(1300, 1400),
            topic="geometry/triangle_congruence",
            solution=steps(
                f"Check if all three sides match (SSS congruence)",
                f"Triangle ABC: sides {side1}, {side2}, {side3}",
                f"Triangle DEF: sides {sides2[0]}, {sides2[1]}, {sides2[2]}",
                f"{'All sides match, triangles are congruent by SSS' if ans else 'Sides do not all match, triangles are not congruent'}",
            ),
            image=d.render(),
        )
    
    elif problem_type == 2:
        # SAS Congruence - given two sides and included angle
        side_a = randint(4, 9)
        side_b = randint(5, 10)
        angle = choice([30, 45, 60, 90, 120])
        
        # Second triangle
        if randint(0, 1) == 0:
            # Congruent
            side_a2, side_b2, angle2 = side_a, side_b, angle
            ans = True
        else:
            # Not congruent - change the angle
            angle2 = angle + choice([-30, -15, 15, 30])
            if angle2 <= 0 or angle2 >= 180:
                angle2 = angle + 30
            side_a2, side_b2 = side_a, side_b
            ans = False
        
        return problem(
            question=f"Triangle ABC has AB = {side_a}, BC = {side_b}, and angle B = {angle}°. Triangle DEF has DE = {side_a2}, EF = {side_b2}, and angle E = {angle2}°. Are the triangles congruent by SAS?",
            answer=ans,
            difficulty=(1400, 1500),
            topic="geometry/triangle_congruence",
            solution=steps(
                f"For SAS congruence, two sides and the included angle must match",
                f"Triangle ABC: sides {side_a}, {side_b} with included angle {angle}°",
                f"Triangle DEF: sides {side_a2}, {side_b2} with included angle {angle2}°",
                f"{'Sides and included angle match, triangles are congruent by SAS' if ans else 'The included angles differ, triangles are not congruent'}",
            ),
        )
    
    elif problem_type == 3:
        # ASA Congruence - given two angles and included side
        angle1 = choice([30, 40, 50, 60])
        angle2 = choice([40, 50, 60, 70])
        while angle1 + angle2 >= 180:
            angle2 = choice([40, 50, 60, 70])
        
        side = randint(5, 12)
        
        # Second triangle
        if randint(0, 1) == 0:
            # Congruent
            angle1_2, angle2_2, side2 = angle1, angle2, side
            ans = True
        else:
            # Not congruent - change one angle
            angle1_2 = angle1 + choice([10, 15, 20])
            angle2_2, side2 = angle2, side
            ans = False
        
        return problem(
            question=f"Triangle ABC has angle A = {angle1}°, angle B = {angle2}°, and side AB = {side}. Triangle DEF has angle D = {angle1_2}°, angle E = {angle2_2}°, and side DE = {side2}. Are the triangles congruent by ASA?",
            answer=ans,
            difficulty=(1400, 1500),
            topic="geometry/triangle_congruence",
            solution=steps(
                f"For ASA congruence, two angles and the included side must match",
                f"Triangle ABC: angles {angle1}° and {angle2}° with included side {side}",
                f"Triangle DEF: angles {angle1_2}° and {angle2_2}° with included side {side2}",
                f"{'Angles and included side match, triangles are congruent by ASA' if ans else 'The angles do not match, triangles are not congruent'}",
            ),
        )
    
    else:
        # Identify congruence theorem from given information
        theorem_choice = randint(1, 3)
        
        if theorem_choice == 1:
            # SSS setup
            s1, s2, s3 = randint(4, 8), randint(5, 9), randint(6, 10)
            while s1 + s2 <= s3 or s1 + s3 <= s2 or s2 + s3 <= s1:
                s3 = randint(6, 10)
            
            return problem(
                question=f"Triangle ABC has sides AB = {s1}, BC = {s2}, AC = {s3}. Triangle DEF has sides DE = {s1}, EF = {s2}, DF = {s3}. Which congruence theorem proves these triangles are congruent?",
                answer="SSS",
                difficulty=(1300, 1400),
                topic="geometry/triangle_congruence",
                solution=steps(
                    f"We are given all three sides of both triangles",
                    f"AB = DE = {s1}, BC = EF = {s2}, AC = DF = {s3}",
                    f"All three pairs of corresponding sides are equal",
                    f"Therefore, the triangles are congruent by SSS (Side-Side-Side)",
                ),
            )
        
        elif theorem_choice == 2:
            # SAS setup
            s1, s2 = randint(5, 10), randint(6, 11)
            ang = choice([45, 60, 90])
            
            return problem(
                question=f"Triangle ABC has AB = {s1}, BC = {s2}, and angle B = {ang}°. Triangle DEF has DE = {s1}, EF = {s2}, and angle E = {ang}°. Which congruence theorem proves these triangles are congruent?",
                answer="SAS",
                difficulty=(1400, 1500),
                topic="geometry/triangle_congruence",
                solution=steps(
                    f"We are given two sides and the included angle of both triangles",
                    f"AB = DE = {s1}, BC = EF = {s2}, and angle B = angle E = {ang}°",
                    f"Two pairs of sides and the included angles are equal",
                    f"Therefore, the triangles are congruent by SAS (Side-Angle-Side)",
                ),
            )
        
        else:
            # ASA setup
            ang1, ang2 = choice([40, 50, 60]), choice([50, 60, 70])
            while ang1 + ang2 >= 180:
                ang2 = choice([50, 60, 70])
            side = randint(7, 12)
            
            return problem(
                question=f"Triangle ABC has angle A = {ang1}°, angle C = {ang2}°, and side AC = {side}. Triangle DEF has angle D = {ang1}°, angle F = {ang2}°, and side DF = {side}. Which congruence theorem proves these triangles are congruent?",
                answer="ASA",
                difficulty=(1400, 1500),
                topic="geometry/triangle_congruence",
                solution=steps(
                    f"We are given two angles and the included side of both triangles",
                    f"Angle A = angle D = {ang1}°, angle C = angle F = {ang2}°, and AC = DF = {side}",
                    f"Two pairs of angles and the included sides are equal",
                    f"Therefore, the triangles are congruent by ASA (Angle-Side-Angle)",
                ),
            )

emit(generate())

File: src/geometry_triangle_congruence_medium.jl
# geometry - triangle_congruence (medium)
# Generated: 2026-03-08T20:03:51.235668

include(joinpath(@__DIR__, "..", "julia", "src", "ProblemUtils.jl"))
using .ProblemUtils

@script begin
    set_topic!("geometry/triangle_congruence")
    
    problem_type = choice(1:4)
    
    if problem_type == 1
        # SSS congruence with diagram
        side1 = randint(5, 12)
        side2 = randint(5, 12)
        side3 = randint(5, 12)
        
        # Ensure triangle inequality
        while side1 + side2 <= side3 || side1 + side3 <= side2 || side2 + side3 <= side1
            side1 = randint(5, 12)
            side2 = randint(5, 12)
            side3 = randint(5, 12)
        end
        
        d = DiagramObj(300, 250)
        
        # Triangle ABC
        ax, ay = 30, 200
        bx, by = 130, 200
        cx, cy = 80, 120
        polygon!(d, [ax, bx, cx], [ay, by, cy], :none, "black", 2)
        point!(d, ax, ay, "A", :below_left)
        point!(d, bx, by, "B", :below_right)
        point!(d, cx, cy, "C", :above)
        segment_label!(d, ax, ay, bx, by, "$(side1)", :below)
        segment_label!(d, bx, by, cx, cy, "$(side2)", :right)
        segment_label!(d, cx, cy, ax, ay, "$(side3)", :left)
        
        # Triangle DEF
        dx, dy = 170, 200
        ex, ey = 270, 200
        fx, fy = 220, 120
        polygon!(d, [dx, ex, fx], [dy, ey, fy], :none, "black", 2)
        point!(d, dx, dy, "D", :below_left)
        point!(d, ex, ey, "E", :below_right)
        point!(d, fx, fy, "F", :above)
        segment_label!(d, dx, dy, ex, ey, "$(side1)", :below)
        segment_label!(d, ex, ey, fx, fy, "$(side2)", :right)
        segment_label!(d, fx, fy, dx, dy, "$(side3)", :left)
        
        image = render(d)
        
        problem(
            question="The diagram shows two triangles. Triangle ABC has sides AB = $(side1), BC = $(side2), and CA = $(side3). Triangle DEF has sides DE = $(side1), EF = $(side2), and FD = $(side3). Which congruence theorem proves that △ABC ≅ △DEF?",
            answer="SSS",
            difficulty=(1200, 1400),
            answer_type="text",
            solution=steps(
                "Given: AB = DE = $(side1), BC = EF = $(side2), CA = FD = $(side3)",
                "All three corresponding sides are congruent",
                sol("Congruence Theorem", "SSS (Side-Side-Side)")
            ),
            image=image,
            time=90
        )
        
    elseif problem_type == 2
        # SAS congruence with diagram
        side1 = randint(6, 14)
        angle = randint(40, 100)
        side2 = randint(6, 14)
        
        d = DiagramObj(300, 250)
        
        # Triangle ABC
        ax, ay = 30, 200
        bx, by = 30 + side1 * 8, 200
        angle_rad = angle * π / 180
        cx = ax + side2 * 8 * cos(angle_rad)
        cy = ay - side2 * 8 * sin(angle_rad)
        polygon!(d, [ax, bx, cx], [ay, by, cy], :none, "black", 2)
        point!(d, ax, ay, "A", :below_left)
        point!(d, bx, by, "B", :below_right)
        point!(d, cx, cy, "C", :above)
        segment_label!(d, ax, ay, bx, by, "$(side1)", :below)
        segment_label!(d, ax, ay, cx, cy, "$(side2)", :left)
        angle_arc!(d, ax, ay, bx, by, cx, cy, "$(angle)°")
        
        # Triangle DEF
        dx, dy = 170, 200
        ex, ey = 170 + side1 * 8, 200
        fx = dx + side2 * 8 * cos(angle_rad)
        fy = dy - side2 * 8 * sin(angle_rad)
        polygon!(d, [dx, ex, fx], [dy, ey, fy], :none, "black", 2)
        point!(d, dx, dy, "D", :below_left)
        point!(d, ex, ey, "E", :below_right)
        point!(d, fx, fy, "F", :above)
        segment_label!(d, dx, dy, ex, ey, "$(side1)", :below)
        segment_label!(d, dx, dy, fx, fy, "$(side2)", :left)
        angle_arc!(d, dx, dy, ex, ey, fx, fy, "$(angle)°")
        
        image = render(d)
        
        problem(
            question="The diagram shows two triangles. In △ABC, AB = $(side1), ∠A = $(angle)°, and AC = $(side2). In △DEF, DE = $(side1), ∠D = $(angle)°, and DF = $(side2). Which congruence theorem proves △ABC ≅ △DEF?",
            answer="SAS",
            difficulty=(1300, 1500),
            answer_type="text",
            solution=steps(
                "Given: AB = DE = $(side1), ∠A = ∠D = $(angle)°, AC = DF = $(side2)",
                "Two sides and the included angle are congruent",
                sol("Congruence Theorem", "SAS (Side-Angle-Side)")
            ),
            image=image,
            time=90
        )
        
    elseif problem_type == 3
        # ASA congruence with diagram
        angle1 = randint(35, 70)
        side = randint(7, 15)
        angle2 = randint(35, 70)
        
        # Ensure valid triangle
        while angle1 + angle2 >= 180
            angle1 = randint(35, 70)
            angle2 = randint(35, 70)
        end
        
        d = DiagramObj(320, 250)
        
        # Triangle ABC
        ax, ay = 30, 200
        bx, by = 30 + side * 8, 200
        angle1_rad = angle1 * π / 180
        angle2_rad = π - angle1_rad - angle2 * π / 180
        cx = bx - 60 * cos(angle2_rad)
        cy = by - 60 * sin(angle2_rad)
        polygon!(d, [ax, bx, cx], [ay, by, cy], :none, "black", 2)
        point!(d, ax, ay, "A", :below_left)
        point!(d, bx, by, "B", :below_right)
        point!(d, cx, cy, "C", :above)
        segment_label!(d, ax, ay, bx, by, "$(side)", :below)
        angle_arc!(d, ax, ay, bx, by, cx, cy, "$(angle1)°")
        angle_arc!(d, bx, by, ax, ay, cx, cy, "$(angle2)°")
        
        # Triangle DEF
        dx, dy = 180, 200
        ex, ey = 180 + side * 8, 200
        fx = ex - 60 * cos(angle2_rad)
        fy = ey - 60 * sin(angle2_rad)
        polygon!(d, [dx, ex, fx], [dy, ey, fy], :none, "black", 2)
        point!(d, dx, dy, "D", :below_left)
        point!(d, ex, ey, "E", :below_right)
        point!(d, fx, fy, "F", :above)
        segment_label!(d, dx, dy, ex, ey, "$(side)", :below)
        angle_arc!(d, dx, dy, ex, ey, fx, fy, "$(angle1)°")
        angle_arc!(d, ex, ey, dx, dy, fx, fy, "$(angle2)°")
        
        image = render(d)
        
        problem(
            question="The diagram shows two triangles. In △ABC, ∠A = $(angle1)°, AB = $(side), and ∠B = $(angle2)°. In △DEF, ∠D = $(angle1)°, DE = $(side), and ∠E = $(angle2)°. Which congruence theorem proves △ABC ≅ △DEF?",
            answer="ASA",
            difficulty=(1300, 1500),
            answer_type="text",
            solution=steps(
                "Given: ∠A = ∠D = $(angle1)°, AB = DE = $(side), ∠B = ∠E = $(angle2)°",
                "Two angles and the included side are congruent",
                sol("Congruence Theorem", "ASA (Angle-Side-Angle)")
            ),
            image=image,
            time=90
        )
        
    else
        # Find missing side/angle using congruence
        side_ab = randint(8, 16)
        side_bc = randint(8, 16)
        side_ca = randint(8, 16)
        
        # Ensure triangle inequality
        while side_ab + side_bc <= side_ca || side_ab + side_ca <= side_bc || side_bc + side_ca <= side_ab
            side_ab = randint(8, 16)
            side_bc = randint(8, 16)
            side_ca = randint(8, 16)
        end
        
        unknown_var = choice(["x", "y"])
        
        d = DiagramObj(300, 250)
        
        # Triangle ABC
        ax, ay = 30, 200
        bx, by = 130, 200
        cx, cy = 80, 120
        polygon!(d, [ax, bx, cx], [ay, by, cy], :none, "black", 2)
        point!(d, ax, ay, "A", :below_left)
        point!(d, bx, by, "B", :below_right)
        point!(d, cx, cy, "C", :above)
        segment_label!(d, ax, ay, bx, by, "$(side_ab)", :below)
        segment_label!(d, bx, by, cx, cy, "$(side_bc)", :right)
        segment_label!(d, cx, cy, ax, ay, "$(side_ca)", :left)
        
        # Triangle DEF with unknown
        dx, dy = 170, 200
        ex, ey = 270, 200
        fx, fy = 220, 120
        polygon!(d, [dx, ex, fx], [dy, ey, fy], :none, "black", 2)
        point!(d, dx, dy, "D", :below_left)
        point!(d, ex, ey, "E", :below_right)
        point!(d, fx, fy, "F", :above)
        segment_label!(d, dx, dy, ex, ey, "$(side_ab)", :below)
        segment_label!(d, ex, ey, fx, fy, "$(unknown_var)", :right)
        segment_label!(d, fx, fy, dx, dy, "$(side_ca)", :left)
        
        image = render(d)
        
        problem(
            question="Given that △ABC ≅ △DEF with AB = $(side_ab), BC = $(side_bc), CA = $(side_ca), DE = $(side_ab), and FD = $(side_ca). The diagram shows both triangles. Find the value of $(unknown_var) where EF = $(unknown_var).",
            answer=side_bc,
            difficulty=(1400, 1600),
            solution=steps(
                "Given: △ABC ≅ △DEF",
                "Corresponding sides of congruent triangles are equal",
                "BC corresponds to EF",
                sol("Therefore, $(unknown_var)", side_bc)
            ),
            image=image,
            time=120
        )
    end
end